File: php/verifyCallback.php
<?php

require_once __DIR__ . '/vendor/autoload.php';
require_once __DIR__ . '/PaylabsSignature.php';

use Dotenv\Dotenv;

$dotenv = Dotenv::createImmutable(__DIR__);
$dotenv->load();

// Performance & SSE Optimization
set_time_limit(0);
ini_set('output_buffering', 'off');
ini_set('zlib.output_compression', false);
while (ob_get_level()) ob_end_flush();
ob_implicit_flush(true);

header('Content-Type: application/json');

$uri = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);
$callbackFile = __DIR__ . '/last_callback.json';

function logEvent($data) {
    global $callbackFile;
    $history = [];
    if (file_exists($callbackFile)) {
        $content = file_get_contents($callbackFile);
        $history = json_decode($content, true);
        if (!is_array($history)) $history = [];
    }
    
    // Add unique ID and timestamp if not present
    if (!isset($data['id'])) {
        $data['id'] = uniqid('ev_', true);
    }
    if (!isset($data['timestamp'])) {
        $data['timestamp'] = date('H:i:s');
    }

    // Check if event already exists for update
    $foundIndex = -1;
    foreach ($history as $index => $item) {
        if ($item['id'] === $data['id']) {
            $foundIndex = $index;
            break;
        }
    }

    if ($foundIndex !== -1) {
        $history[$foundIndex] = array_merge($history[$foundIndex], $data);
    } else {
        // Prepend new event
        array_unshift($history, $data);
        // Limit to 50 items
        $history = array_slice($history, 0, 50);
    }
    
    file_put_contents($callbackFile, json_encode($history, JSON_PRETTY_PRINT));
}

// Serve Frontend
if ($uri === '/' || $uri === '/index.html') {
    header('Content-Type: text/html');
    echo file_get_contents(__DIR__ . '/views/index.html');
    exit;
}

// Serve CSS
if ($uri === '/css/style.css') {
    header('Content-Type: text/css');
    echo file_get_contents(__DIR__ . '/views/css/style.css');
    exit;
}

// Serve JS
if ($uri === '/js/client.js') {
    header('Content-Type: application/javascript');
    echo file_get_contents(__DIR__ . '/views/js/client.js');
    exit;
}

// One-shot SSE Endpoint
if ($uri === '/events') {
    header('Content-Type: text/event-stream');
    header('Cache-Control: no-cache');
    header('Connection: keep-alive');

    if (file_exists($callbackFile)) {
        $content = file_get_contents($callbackFile);
        $history = json_decode($content, true);
        if (is_array($history)) {
            // Send history in reverse (oldest first) so client appends correctly
            foreach (array_reverse($history) as $data) {
                echo "data: " . json_encode($data) . "\n\n";
            }
        }
    }
    // End the connection immediately after history delivery to avoid blocking
    // the single-threaded PHP built-in server.
    exit;
}

// Outbound Log Endpoint (from generateTransaction.php)
if ($uri === '/log' && $_SERVER['REQUEST_METHOD'] === 'POST') {
    $rawBody = file_get_contents('php://input');
    $logData = json_decode($rawBody, true);
    if ($logData) {
        logEvent($logData);
    }
    echo json_encode(['status' => 'success']);
    exit;
}

// Handle Callback
if ($uri === '/callback' && $_SERVER['REQUEST_METHOD'] === 'POST') {
    // Get headers
    $signature = $_SERVER['HTTP_X_SIGNATURE'] ?? '';
    $timestamp = $_SERVER['HTTP_X_TIMESTAMP'] ?? '';
    $publicKey = $_ENV['PAYLABS_PUBLIC_KEY'];
    
    error_log("Incoming Callback Headers:");
    foreach (getallheaders() as $name => $value) {
        error_log("  {$name}: {$value}");
    }
    
    // Read raw body
    $rawBody = file_get_contents('php://input');
    
    // Calculate hash
    $shaJson = strtolower(hash('sha256', $rawBody));
    
    // Build string to verify
    $stringToVerify = "POST:/callback:{$shaJson}:{$timestamp}";
    error_log("String to Verify: {$stringToVerify}");
    
    // Verify signature
    $valid = PaylabsSignature::verifySignature($stringToVerify, $signature, $publicKey);
    
    $body = json_decode($rawBody, true);
    
    $requestId = $body['requestId'] ?? '';
    $merchantId = $body['merchantId'] ?? '';
    $status = $body['status'] ?? '';

    $responseData = $status !== '02' ? [
        'requestId' => $requestId,
        'errCode' => '1',
        'errCodeDes' => 'Payment not completed',
        'merchantId' => $merchantId,
    ] : [
        'requestId' => $requestId,
        'errCode' => '0',
        'errCodeDes' => 'Success',
        'merchantId' => $merchantId,
    ];

    // Broadcast via file for SSE
    logEvent([
        'id' => $requestId ?: uniqid('ev_', true),
        'type' => 'inbound',
        'headers' => getallheaders(),
        'body' => $body,
        'endpoint' => '/callback',
        'verificationStatus' => $valid ? 'Valid' : 'Invalid',
        'responseBody' => $responseData
    ]);
    
    if (!$valid) {
        http_response_code(400);
        echo json_encode(['error' => 'Invalid Signature']);
        exit;
    }
    
    error_log("Signature is valid");
    
    echo json_encode($responseData);
    exit;
}

// Handle SNAP Callback
if ($uri === '/transfer-va/payment' && $_SERVER['REQUEST_METHOD'] === 'POST') {
    // Get headers
    $signature = $_SERVER['HTTP_X_SIGNATURE'] ?? '';
    $timestamp = $_SERVER['HTTP_X_TIMESTAMP'] ?? '';
    $partnerId = $_SERVER['HTTP_X_PARTNER_ID'] ?? '';
    $externalId = $_SERVER['HTTP_X_EXTERNAL_ID'] ?? '';
    $ipAddress = $_SERVER['HTTP_X_IP_ADDRESS'] ?? '';
    $publicKey = $_ENV['PAYLABS_PUBLIC_KEY'];
    
    error_log("Incoming SNAP Callback Headers:");
    foreach (getallheaders() as $name => $value) {
        error_log("  {$name}: {$value}");
    }
    
    // Read raw body
    $rawBody = file_get_contents('php://input');
    
    // Calculate hash
    $shaJson = strtolower(hash('sha256', $rawBody));
    
    // Build string to verify
    $stringToVerify = "POST:/transfer-va/payment:{$shaJson}:{$timestamp}";
    error_log("SNAP String to Verify: {$stringToVerify}");
    
    // Verify signature
    $valid = PaylabsSignature::verifySignature($stringToVerify, $signature, $publicKey);
    
    $body = json_decode($rawBody, true);
    
    // Filter allowed fields
    $allowedFields = [
        "paidBills", "virtualAccountNo", "paymentRequestId", "partnerServiceId",
        "virtualAccountPhone", "virtualAccountName", "journalNum", "flagAdvise",
        "trxId", "paymentFlagReason", "virtualAccountEmail", "billDetails",
        "totalAmount", "customerNo", "paymentType", "paidAmount", "referenceNo",
        "trxDateTime", "freeTexts", "paymentFlagStatus"
    ];
    
    $filteredBody = [];
    foreach ($allowedFields as $field) {
        if (array_key_exists($field, $body)) {
            $filteredBody[$field] = $body[$field];
        }
    }
    
    // paymentFlagStatus inside virtualAccountData
    $filteredBody['paymentFlagStatus'] = "00";
    
    $responseData = [
        'responseCode' => '2002500',
        'responseMessage' => 'Success',
        'virtualAccountData' => $filteredBody,
    ];

    $paymentRequestId = $body['paymentRequestId'] ?? '';

    // Broadcast via file for SSE
    logEvent([
        'id' => $paymentRequestId ?: uniqid('ev_', true),
        'type' => 'inbound',
        'headers' => getallheaders(),
        'body' => $body,
        'endpoint' => '/transfer-va/payment',
        'verificationStatus' => $valid ? 'Valid' : 'Invalid',
        'responseBody' => $responseData
    ]);
    
    if (!$valid) {
        http_response_code(401);
        echo json_encode(['responseCode' => '4010000', 'responseMessage' => 'Unauthorized']);
        exit;
    }
    
    error_log("SNAP Signature is valid");

    echo json_encode($responseData);
    exit;
}

// Handle SNAP Create VA Callback
if ($uri === '/api/v1.0/transfer-va/create-va' && $_SERVER['REQUEST_METHOD'] === 'POST') {
    $signature = $_SERVER['HTTP_X_SIGNATURE'] ?? '';
    $timestamp = $_SERVER['HTTP_X_TIMESTAMP'] ?? '';
    $publicKey = $_ENV['PAYLABS_PUBLIC_KEY'];

    error_log("Incoming SNAP Create VA Headers:");
    foreach (getallheaders() as $name => $value) {
        error_log("  {$name}: {$value}");
    }

    $rawBody = file_get_contents('php://input');
    $shaJson = strtolower(hash('sha256', $rawBody));

    $stringToVerify = "POST:/transfer-va/create-va:{$shaJson}:{$timestamp}";
    error_log("SNAP Create VA String to Verify: {$stringToVerify}");

    $valid = PaylabsSignature::verifySignature($stringToVerify, $signature, $publicKey);
    $body = json_decode($rawBody, true);

    $responseCode = $valid ? "2002700" : "4012701";
    $responseMessage = $valid ? "Success" : "Invalid Signature";

    $responseData = [
        'responseCode' => $responseCode,
        'responseMessage' => $responseMessage
    ];

    $paymentRequestId = $body['paymentRequestId'] ?? '';
    logEvent([
        'id' => $paymentRequestId ?: uniqid('ev_', true),
        'type' => 'inbound',
        'headers' => getallheaders(),
        'body' => $body,
        'endpoint' => '/api/v1.0/transfer-va/create-va',
        'verificationStatus' => $valid ? 'Valid' : 'Invalid',
        'responseBody' => $responseData
    ]);

    if (!$valid) {
        http_response_code(401);
    }
    echo json_encode($responseData);
    exit;
}

// 404 for other routes
if ($uri !== '/') {
    http_response_code(404);
    echo json_encode(['error' => 'Not Found']);
    exit;
}

File: dotnet/VerifyCallback.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Microsoft.AspNetCore.Mvc;

namespace QuickStart;

public static class VerifyCallback
{
    private static readonly List<StreamWriter> Clients = new();
    private static readonly SemaphoreSlim ClientsLock = new(1, 1);

    public static void MapCallbackEndpoint(WebApplication app)
    {
        app.MapGet("/events", async (HttpContext context) =>
        {
            context.Response.Headers.Add("Content-Type", "text/event-stream");
            context.Response.Headers.Add("Cache-Control", "no-cache");
            context.Response.Headers.Add("Connection", "keep-alive");

            var writer = new StreamWriter(context.Response.Body);
            await ClientsLock.WaitAsync();
            try
            {
                Clients.Add(writer);
            }
            finally
            {
                ClientsLock.Release();
            }

            try
            {
                while (!context.RequestAborted.IsCancellationRequested)
                {
                    await Task.Delay(1000);
                }
            }
            finally
            {
                await ClientsLock.WaitAsync();
                try
                {
                    Clients.Remove(writer);
                }
                finally
                {
                    ClientsLock.Release();
                }
            }
        });

        app.MapPost("/callback", async (HttpContext context) =>
        {
            var signature = context.Request.Headers["X-Signature"].FirstOrDefault() ?? "";
            var timestamp = context.Request.Headers["X-Timestamp"].FirstOrDefault() ?? "";
            var publicKey = Environment.GetEnvironmentVariable("PAYLABS_PUBLIC_KEY") ?? "";

            Console.WriteLine("Incoming Callback Headers:");
            var headers = new Dictionary<string, string>();
            foreach (var header in context.Request.Headers)
            {
                Console.WriteLine($"  {header.Key}: {header.Value}");
                headers[header.Key.ToLower()] = header.Value.ToString();
            }

            using var reader = new StreamReader(context.Request.Body);
            var rawBody = await reader.ReadToEndAsync();

            using var sha256 = SHA256.Create();
            var bodyHash = Convert.ToHexString(sha256.ComputeHash(Encoding.UTF8.GetBytes(rawBody))).ToLower();

            var stringToVerify = $"POST:/callback:{bodyHash}:{timestamp}";
            Console.WriteLine($"String to Verify: {stringToVerify}");

            var valid = PaylabsSignature.VerifySignature(stringToVerify, signature, publicKey);

            // Broadcast
            var bodyElement = JsonDocument.Parse(rawBody).RootElement;
            var jsonBody = JsonSerializer.Deserialize<object>(rawBody); // For simple object structure

            var sseData = new
            {
                headers,
                body = jsonBody,
                verificationStatus = valid ? "Valid" : "Invalid"
            };
            
            var sseMessage = $"data: {JsonSerializer.Serialize(sseData)}\n\n";

            await ClientsLock.WaitAsync();
            try
            {
                foreach (var client in Clients)
                {
                    try
                    {
                        await client.WriteAsync(sseMessage);
                        await client.FlushAsync();
                    }
                    catch
                    {
                        // Handle disconnected clients
                    }
                }
            }
            finally
            {
                ClientsLock.Release();
            }

            var status = bodyElement.TryGetProperty("status", out var s) ? s.GetString() : "";
            var requestId = bodyElement.TryGetProperty("requestId", out var r) ? r.GetString() : "";
            var merchantId = bodyElement.TryGetProperty("merchantId", out var m) ? m.GetString() : "";

            var responseData = status != "02" ? 
                (object)new
                {
                    requestId,
                    errCode = "1",
                    errCodeDes = "Payment not completed",
                    merchantId
                } : 
                new
                {
                    requestId,
                    errCode = "0",
                    errCodeDes = "Success",
                    merchantId
                };

            // Capture and broadcast again with responseBody
            var finalSseData = new
            {
                type = "inbound",
                headers,
                body = jsonBody,
                endpoint = "/callback",
                verificationStatus = valid ? "Valid" : "Invalid",
                responseBody = responseData
            };

            await BroadcastSse(JsonSerializer.Serialize(finalSseData));

            return Results.Ok(responseData);
        });

        app.MapPost("/api/v1.0/transfer-va/create-va", async (HttpContext context) =>
        {
            var signature = context.Request.Headers["X-Signature"].FirstOrDefault() ?? "";
            var timestamp = context.Request.Headers["X-Timestamp"].FirstOrDefault() ?? "";
            var publicKey = Environment.GetEnvironmentVariable("PAYLABS_PUBLIC_KEY") ?? "";

            Console.WriteLine("Incoming SNAP Create VA Headers:");
            var headers = new Dictionary<string, string>();
            foreach (var header in context.Request.Headers)
            {
                Console.WriteLine($"  {header.Key}: {header.Value}");
                headers[header.Key.ToLower()] = header.Value.ToString();
            }

            using var reader = new StreamReader(context.Request.Body);
            var rawBody = await reader.ReadToEndAsync();

            using var sha256 = SHA256.Create();
            var bodyHash = Convert.ToHexString(sha256.ComputeHash(Encoding.UTF8.GetBytes(rawBody))).ToLower();

            // Pattern: POST:/transfer-va/create-va:{bodyHash}:{timestamp}
            var stringToVerify = $"POST:/transfer-va/create-va:{bodyHash}:{timestamp}";
            Console.WriteLine($"SNAP Create VA String to Verify: {stringToVerify}");

            var valid = PaylabsSignature.VerifySignature(stringToVerify, signature, publicKey);
            var jsonBody = JsonSerializer.Deserialize<object>(rawBody);

            var responseCode = valid ? "2002700" : "4012701";
            var responseMessage = valid ? "Success" : "Invalid Signature";

            var responseData = new
            {
                responseCode,
                responseMessage
            };

            // Broadcast
            var sseData = new
            {
                type = "inbound",
                headers,
                body = jsonBody,
                endpoint = "/api/v1.0/transfer-va/create-va",
                verificationStatus = valid ? "Valid" : "Invalid",
                responseBody = responseData
            };
            
            await BroadcastSse(JsonSerializer.Serialize(sseData));

            if (!valid)
            {
                context.Response.StatusCode = 401;
                await context.Response.WriteAsJsonAsync(responseData);
                return;
            }

            Console.WriteLine("SNAP Create VA Signature is valid");
            await context.Response.WriteAsJsonAsync(responseData);
        });

        app.MapPost("/log", async (HttpContext context) =>
        {
            using var reader = new StreamReader(context.Request.Body);
            var rawBody = await reader.ReadToEndAsync();
            await BroadcastSse(rawBody);
            return Results.Ok();
        });

        app.MapPost("/transfer-va/payment", async (HttpContext context) =>
        {
            var signature = context.Request.Headers["X-Signature"].FirstOrDefault() ?? "";
            var timestamp = context.Request.Headers["X-Timestamp"].FirstOrDefault() ?? "";
            var publicKey = Environment.GetEnvironmentVariable("PAYLABS_PUBLIC_KEY") ?? "";

            Console.WriteLine("Incoming SNAP Callback Headers:");
            var headers = new Dictionary<string, string>();
            foreach (var header in context.Request.Headers)
            {
                Console.WriteLine($"  {header.Key}: {header.Value}");
                headers[header.Key.ToLower()] = header.Value.ToString();
            }

            using var reader = new StreamReader(context.Request.Body);
            var rawBody = await reader.ReadToEndAsync();

            using var sha256 = SHA256.Create();
            var bodyHash = Convert.ToHexString(sha256.ComputeHash(Encoding.UTF8.GetBytes(rawBody))).ToLower();

            var stringToVerify = $"POST:/transfer-va/payment:{bodyHash}:{timestamp}";
            Console.WriteLine($"SNAP String to Verify: {stringToVerify}");

            var valid = PaylabsSignature.VerifySignature(stringToVerify, signature, publicKey);

            // Broadcast
            var bodyElement = JsonDocument.Parse(rawBody).RootElement;
            var jsonBody = JsonSerializer.Deserialize<object>(rawBody);

            var sseData = new
            {
                headers,
                body = jsonBody,
                endpoint = "/transfer-va/payment",
                verificationStatus = valid ? "Valid" : "Invalid"
            };
            
            var sseMessage = $"data: {JsonSerializer.Serialize(sseData)}\n\n";

            await ClientsLock.WaitAsync();
            try
            {
                foreach (var client in Clients)
                {
                    try
                    {
                        await client.WriteAsync(sseMessage);
                        await client.FlushAsync();
                    }
                    catch
                    {
                        // Handle disconnected clients
                    }
                }
            }
            finally
            {
                ClientsLock.Release();
            }

            if (!valid)
            {
                context.Response.StatusCode = 401;
                await context.Response.WriteAsJsonAsync(new { responseCode = "4010000", responseMessage = "Unauthorized" });
                return;
            }

            Console.WriteLine("SNAP Signature is valid");

            var allowedFields = new[] {
                "paidBills", "virtualAccountNo", "paymentRequestId", "partnerServiceId",
                "virtualAccountPhone", "virtualAccountName", "journalNum", "flagAdvise",
                "trxId", "paymentFlagReason", "virtualAccountEmail", "billDetails",
                "totalAmount", "customerNo", "paymentType", "paidAmount", "referenceNo",
                "trxDateTime", "freeTexts", "paymentFlagStatus"
            };

            var filteredBody = new Dictionary<string, object>();
            foreach (var property in bodyElement.EnumerateObject())
            {
                if (allowedFields.Contains(property.Name))
                {
                    // Basic deserialization of property value
                    filteredBody[property.Name] = JsonSerializer.Deserialize<object>(property.Value.GetRawText());
                }
            }
            
            // paymentFlagStatus inside virtualAccountData
            filteredBody["paymentFlagStatus"] = "00";

            var responseData = new
            {
                responseCode = "2002500",
                responseMessage = "Success",
                virtualAccountData = filteredBody
            };

            // Update SSE with responseBody
            var finalSseData = new
            {
                type = "inbound",
                headers,
                body = jsonBody,
                endpoint = "/transfer-va/payment",
                verificationStatus = "Valid",
                responseBody = responseData
            };
            await BroadcastSse(JsonSerializer.Serialize(finalSseData));

            await context.Response.WriteAsJsonAsync(responseData);
        });
    }

    private static async Task BroadcastSse(string message)
    {
        var sseMessage = $"data: {message}\n\n";

        await ClientsLock.WaitAsync();
        try
        {
            foreach (var client in Clients)
            {
                try
                {
                    await client.WriteAsync(sseMessage);
                    await client.FlushAsync();
                }
                catch
                {
                    // Handle disconnected clients
                }
            }
        }
        finally
        {
            ClientsLock.Release();
        }
    }
}
